src/tactic/congrm.lean

```diff
@@ -0,0 +1,134 @@
+
+`g` is possibly the proof of an equality?  It is extracted from the first `list expr`.
+-/
+private meta def extract_subgoals : list expr → list congr_arg_kind → list expr →
+  tactic (list (expr × expr))
+| (_ :: _ :: g :: prf_args) (congr_arg_kind.eq :: kinds)             (pat :: pat_args) :=
+  (λ rest, (g, pat) :: rest) <$> extract_subgoals prf_args kinds pat_args
+| (_ :: prf_args)           (congr_arg_kind.fixed :: kinds)          (_ :: pat_args) :=
+  extract_subgoals prf_args kinds pat_args
+| prf_args                  (congr_arg_kind.fixed_no_param :: kinds) (_ :: pat_args) :=
+  extract_subgoals prf_args kinds pat_args
+| (_ :: _ :: prf_args)      (congr_arg_kind.cast :: kinds)           (_ :: pat_args) :=
+  extract_subgoals prf_args kinds pat_args
+| _ _ [] := pure []
+| _ _ _ := fail "unsupported congr lemma"
+
+/--
+`equate_with_pattern_core pat` solves a single goal of the form `lhs = rhs`
+(assuming that `lhs` and `rhs` are unifiable with `pat`)
+by applying congruence lemmas until `pat` is a metavariable.
+Returns the list of metavariables for the new subgoals at the leafs.
+Calls `set_goals []` at the end.
+-/
+meta def equate_with_pattern_core : expr → tactic (list expr) | pat :=
+(applyc ``subsingleton.elim >> pure []) <|>
+(applyc ``rfl >> pure []) <|>
+if pat.is_mvar || pat.get_delayed_abstraction_locals.is_some then do
+  try $ applyc ``_root_.propext,
+  get_goals <* set_goals []
+else if pat.is_app then do
+  cl ← mk_specialized_congr_lemma pat,
+  H_congr_lemma ← assertv `H_congr_lemma cl.type cl.proof,
+  [prf] ← get_goals,
+  apply H_congr_lemma <|> fail "could not apply congr_lemma",
+  all_goals' $ try $ clear H_congr_lemma,  -- given the `set_goals []` that follows, is this needed?
+  set_goals [],
+  prf ← instantiate_mvars prf,
+  subgoals ← extract_subgoals prf.get_app_args cl.arg_kinds pat.get_app_args,
+  subgoals ← subgoals.mmap (λ ⟨subgoal, subpat⟩, do
+    set_goals [subgoal],
+    equate_with_pattern_core subpat),
+  pure subgoals.join
+else if pat.is_lambda then do
+  applyc ``_root_.funext,
+  x ← intro pat.binding_name,
+  equate_with_pattern_core $ pat.lambda_body.instantiate_var x
+else if pat.is_pi then do
+  applyc ``_root_.pi_congr,
+  x ← intro pat.binding_name,
+  equate_with_pattern_core $ pat.pi_codomain.instantiate_var x
+else do
+  pat ← pp pat,
+  fail $ to_fmt "unsupported pattern:\n" ++ pat
+
+/--
+`equate_with_pattern pat` solves a single goal of the form `lhs = rhs`
+(assuming that `lhs` and `rhs` are unifiable with `pat`)
+by applying congruence lemmas until `pat` is a metavariable.
+The subgoals for the leafs are prepended to the goals.
+--/
+meta def equate_with_pattern (pat : expr) : tactic unit := do
+congr_subgoals ← solve1 (equate_with_pattern_core pat),
+gs ← get_goals,
+set_goals $ congr_subgoals ++ gs
+
+end tactic
+
+namespace tactic.interactive
+open tactic interactive
+setup_tactic_parser
+
+/--
+`congrm e` assumes that the goal is of the form `lhs = rhs` or `lhs ↔ rhs`.
```

Comment:
My only complaint is that the docs are a little on the technical side, with discussion of metavariables and so on! I think it might be clearer to write something like, "`congrm e` takes an expression `e` containing placeholders `_`. It matches both `lhs` and `rhs` to this pattern, and produces one goal for each placeholder, stating that the corresponding subexpressions in `lhs` and `rhs` are equal."

I think it would also be good to include one or two of the earlier tests in the docs as examples, rather than only giving the one that has binders in it.

Reply:
Thanks, Rob!  I updated the doc-string, making it hopefully more user-friendly!

I also added an example that should clarify better the intended use of the tactic.